Feature/#12 서비스별 패키지 나누기

chanwooo opened the pull request:
- 도메인별로 패키지 만들기
- 각 도메인별 서비스 인터페이스 작성

reasonSun committed:
test/#8_대기열 테스트 작성

chanwooo committed:
test/#8_MakeDummy
dummy 만드는 클래스 추가

chanwooo committed:
refactor/#12_도메인별 패키지 추가, 서비스 레이어 생성

chanwooo committed:
refactor/#12_임시저장

chanwooo committed:
refactor/#12_waitingQueue -> waiting으로 이름 변경

chanwooo committed:
refactor/#12_큐잉서비스 request, response 작성

chanwooo committed:
refactor/#12_review, rating - request, response 작성

chanwooo committed:
minor/type변경, 변수이름수정

reasonSun committed:
refactor/#12_menu,store,user - request,response 작성

reasonSun committed:
Merge remote-tracking branch 'origin/feature/#12_서비스별_패키지_나누기' into feature/#12_서비스별_패키지_나누기

f-lab-trey commented on src/main/java/flab/quing/review/RatingService.java:
```diff
@@ -0,0 +1,14 @@
+
+    RatingResponse getRating(Long storeId);
+
+    /**
+     * 모든 상점의 리뷰와 별점을 계산해서 각 상점별 별점을 갱신합니다.
+     */
+    void updateRating();
```
Self-descriptive하게 메소드 이름을 변경해보는 게 어떨까요?

chanwooo replied:
.updateAll() 정도로 설명이 될거같기도하네요.

f-lab-trey commented on src/main/java/flab/quing/review/dto/RatingResponse.java:
```diff
@@ -0,0 +1,8 @@
+package flab.quing.review.dto;
+
+public class RatingResponse {
```
[질문]
lombok annotation이 안 붙어있어서 default creator만 있는 상태일텐데, 이 때 ObjectMapper 등으로 serialize-deserialize가 가능한가요?

chanwooo replied:
getter가 필요하겠네요.
service에서 return하면서 급하게 만들고 추후에 서비스코드짜면서 수정할생각에 어노테이션을 달지 않았어요.
 @getter 정도 붙이면 간단히 해결될거같은데 수정하겠습니다~

f-lab-trey commented on src/main/java/flab/quing/review/dto/ReviewRequest.java:
```diff
@@ -0,0 +1,10 @@
+package flab.quing.review.dto;
+
+public class ReviewRequest {
```
[위 comment에 이어서]
Default creator만 가지고는 이 클래스를 테스트 코드에서 이용할 수 없을 거 같아요.

f-lab-trey commented on src/main/java/flab/quing/review/dto/RatingResponse.java:
```diff
@@ -0,0 +1,8 @@
+
+public class RatingResponse {
+
+    private Long storeId;
```
`storeId`가 nullable인가요? 아니라면 Boxed 형태를 사용하지 않아도 괜찮을 거 같습니다. 
다른 field에도 같은 의견입니다.

chanwooo replied:
다른예제에서 id를 받는경우 Long을 쓰는경우가 많아보여서 nullable이 필요한곳도있을거같다는생각에 boxed된 변수로 통일을 하였는데.. 필요한곳에만 적용해야겠네요.

chanwooo replied:
id를 담는경우에 id가 0인지 값이 없어서 0인지 구분하기 어렵기때문에 wrapper type을 사용한다고 하는데 일리가 있다고 생각해요.

https://akdl911215.tistory.com/299

f-lab-trey commented on src/main/java/flab/quing/store/Menu.java:
```diff
@@ -10,7 +11,7 @@
 @Getter
 @Setter
-public class Menu extends BaseEntity{
+public class Menu extends BaseEntity {
```
[모든 Entity에 적용]
`@Data` annotation을 사용하지 않았기 때문에, `hashCode()`나 `equals()`가 정의되어 있지 않은 상태인데요, 이럴 경우 entity들이 map 등의 collection에서 사용될 때 문제가 발생할 수 있을 거 같습니다. 관련하여 찾아보시고 멘토링 시간에 같이 얘기해봐요. 참고로, 예전에 카카오페이 면접때 이 질문을 받았던 기억이 나네요.

chanwooo replied:
`@Data`를 안쓰자니 너무 주렁주렁 달게되고.. hashcode and equals가 정의안되면 인스턴스id로 중복을 체크할테니 
내용은 같지만 따로 생성한 객체의경우 다른것으로 판단하게될테니 hashcode, equlas를 정의해야하는거라 알고있는데

chanwooo replied:
어노테이션도 주렁주렁 달고싶지않고 setter도 쓰지못하게 하고싶던와중 훌륭한 게비스콘을 찾았네요..
```java
@Data
@Setter(AccessLevel.NONE)
```
https://velog.io/@bwjhj1030/DTO-%EB%A7%8C%EB%93%A4-%EB%95%8C-Lombok-%EA%BF%80%ED%8C%81-%EB%8C%80%EB%B0%A9%EC%B6%9C

reasonSun replied:
찬우 님이 찾으신 자료 참고하여 어노테이션 적용 했습니다.

f-lab-trey commented on src/main/java/flab/quing/store/dto/MenuRequest.java:
```diff
@@ -0,0 +1,18 @@
+import lombok.Getter;
+import lombok.NoArgsConstructor;
+
+@Getter
+@NoArgsConstructor
+@Builder
+@AllArgsConstructor
```
`@Data`로 대체하면 더 편하지 않을까요?

chanwooo replied:
setter를 만들고싶지않은마음에서 저렇게 썼던거같아요. 고민되는부분이네요..

reasonSun replied:
넵 아래와 같이 수정 했습니다!
```
@Data
@Setter(AccessLevel.NONE)
@Builder
```

f-lab-trey commented on src/main/java/flab/quing/store/dto/MenuRequest.java:
```diff
@@ -0,0 +1,18 @@
+public class MenuRequest {
+
+    private String name;
+    private String image;
```
`imageURL` 또는 `imageUrl` 등이 더 적절해보입니다.

reasonSun replied:
네. `imageUrl` 로 변경했습니다!

f-lab-trey commented on src/main/java/flab/quing/store/dto/MenuResponse.java:
```diff
@@ -0,0 +1,19 @@
+    private String image;
+    private Integer price;
+
+    public MenuResponse(Long menuId, String name, String image, Integer price) {
+        this.menuId = menuId;
+        this.name = name;
+        this.image = image;
+        this.price = price;
+    }
```
`@AllArgsConstructor`를 사용하면 이 구현을 생략할 수 있을 거 같습니다.

chanwooo replied:
이부분은 유선님이랑 둘이서 리뷰하면서도 얘기했던부분이라 몇군데는 지웠는데 빠뜨린것도 있네요.ㅎㅎ

reasonSun replied:
수정했습니다!

f-lab-trey commented on src/main/java/flab/quing/store/dto/StoreResponse.java:
```diff
@@ -0,0 +1,25 @@
+    private String address;
+    private String pageLink;
+
+    public StoreResponse(Long id, String name, String phoneNumber, String openStatus, String openHours, String address, String pageLink) {
```
ditto

f-lab-trey commented on src/main/java/flab/quing/user/MemberRepository.java:
```diff
@@ -0,0 +1,11 @@
+
+public interface MemberRepository extends JpaRepository<User, Long> {
+    Optional<User> findByName(String name);
+    Optional<User> findByPhone(String phone);
```
`phoneNumber`로 변경해주세요.

f-lab-trey commented on src/main/java/flab/quing/user/dto/StoreManagerResponse.java:
```diff
@@ -0,0 +1,19 @@
+    private String phoneNumber;
+    private String name;
+
+    public StoreManagerResponse(String loginId, String password, String phoneNumber, String name) {
```
ditto

f-lab-trey commented on src/main/java/flab/quing/user/dto/UserResponse.java:
```diff
@@ -0,0 +1,17 @@
+    private String phoneNumber;
+    private String name;
+
+    public UserResponse(Long userId, String phoneNumber, String name) {
```
ditto

f-lab-trey commented on src/main/java/flab/quing/waiting/QuingService.java:
```diff
@@ -0,0 +1,25 @@
+
+    Integer countForward(Long waitingId);
+
+    void sendMessage(Long waitingId, String msg);
```
[minor]
저는 `message`로 변경하는 게 더 좋을 거 같은데요, `msg`로도 많이 사용해서 취향에 따라 선택하면 될 거 같네요.

chanwooo replied:
넵 비밀번호를 pw라고 썼던거에 우린 자바를 하고있다고하셨던 말씀이 다시한번 떠오르네요..ㅋㅋㅋ ㅠㅠ

f-lab-trey commented on src/main/java/flab/quing/review/Review.java:
```diff
@@ -10,9 +13,9 @@
+public class Review extends BaseEntity {
     @ManyToOne(cascade = CascadeType.ALL)
-    private Member member;
+    private User member;
```
`user`로 통일하는 것이 어떨까요? 여기는 `member`인데 다른 클래스를 보면 `userName`등으로 사용하고 있네요.

chanwooo replied:
Member에서 User로 이름을 바꾸면서 리팩토링 기능을 사용하였는데 미쳐 신경쓰지못한부분이네요.

f-lab-trey commented on src/main/java/flab/quing/waiting/WaitingQueueService.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public Waiting popWaitingQueue(Store store, User member) {
+
+        return null;
```
`null`을 리턴하는 대신 not implemented 관련 exception을 던지는 것으로 의미를 더 확실히 전달할 수 있습니다.

chanwooo replied:
```java
 throw new UnsupportedOperationException("Not Implemented, yet");
 ```
이런식으로 대체하였습니다.

f-lab-trey commented on src/main/java/flab/quing/waiting/dto/WaitingRequest.java:
```diff
@@ -0,0 +1,8 @@
+
+public class WaitingRequest {
+
+    private Long userId = -1L;
```
이 default 값은 필요한가요?

chanwooo replied:
`phoneNumber`랑 `userId`가 둘중에 하나만 들어오면 한쪽은 필요없어질거라 생각해서 
기본값을 적용하면 어떤 해결책이 되지않을까라고 생각했는데 require가 아니니 그냥 비워둬도 상관없겠네요. 삭제하겠습니다~

f-lab-trey commented on src/main/java/flab/quing/waiting/dto/WaitingRequest.java:
```diff
@@ -0,0 +1,8 @@
+public class WaitingRequest {
+
+    private Long userId = -1L;
+    private String phoneNumber = "";
```
ditto

f-lab-trey commented on src/test/java/flab/quing/repository/StoreRatingRepositoryTest.java:
```diff
@@ -31,36 +33,36 @@ class StoreRatingRepositoryTest {
-             storeReviews) {
+        List<Review> storeReviews = storeReviewRepository.findAllByStore(store);
+        Float rating = 0f;
+        for (Review review :
+                storeReviews) {
```
불필요한 newline 이라고 생각합니다.

f-lab-trey commented on src/test/java/flab/quing/repository/StoreReviewRepositoryTest.java:
```diff
@@ -28,12 +28,12 @@ class StoreReviewRepositoryTest {
         store.setPageLink("http://test.co.kr");
 
-        Member member = new Member();
+        User member = new User();
```
ditto

f-lab-trey commented on src/test/java/flab/quing/repository/WaitingQueueRepositoryTest.java:
```diff
@@ -23,15 +24,15 @@ class WaitingQueueRepositoryTest {
         store.setName("큐잉상점");
 
-        Member member1 = new Member();
+        User member1 = new User();
```
ditto

f-lab-trey commented on src/test/java/flab/quing/repository/WaitingQueueRepositoryTest.java:
```diff
@@ -42,16 +43,16 @@ class WaitingQueueRepositoryTest {
         store.setName("큐잉상점");
 
-        Member member1 = new Member();
+        User member1 = new User();
```
ditto

f-lab-trey commented on src/test/java/flab/quing/service/WaitingQueueServiceTest.java:
```diff
@@ -0,0 +1,79 @@
+
+    @Test
+    void 대기열등록_성공() {
+        waitingQueueService = new WaitingQueueService(waitingQueueRepository);
```
위에서 `@InjectMocks` 했기 때문에 여기서 service의 instance를 새로 만드는 것은 중복입니다.

f-lab-trey commented on src/test/java/flab/quing/service/WaitingQueueServiceTest.java:
```diff
@@ -0,0 +1,79 @@
+        Assertions.assertThat(pushedItem.getMember().getName()).isEqualTo(member.getName());
+
+    }
+    static class MakeDummy {
```
1. 이름을 `DummyDataMaker` 정도로 변경하는 것이 더 좋을 거 같습니다.
2. 다른 테스트에서도 사용할 수 있다면, 유틸리티 클래스로 따로 빼도 괜찮을 거 같습니다.

f-lab-trey commented:
제 comment를 모두 반영하지 않으셔도 됩니다. 관련하여 다른 의견이 있으시면 thread에 편하게 의견 남겨주세요. 👍

chanwooo commented:
음.. 일단 test관련 코드들을 포함한,, service interface 이외의 코드들은 프로젝트 설계 이전에 작성한것들이에요.
그부분을 가리고 리뷰요청을 하고싶었는데 그부분을 잘 모르겠어서 그냥 모든수정사항이 모두 들어있는채로 올라가게되었네요.ㅠㅠ
test쪽은 코드가 바뀐게 많아서 아마 다시짜게될거같아요. 리뷰 반영해서 커밋하도록하겠습니다~!

chanwooo committed:
review반영/self-descriptive한 method명 updateRating->updateAll()

chanwooo committed:
review반영/msg -> message

chanwooo committed:
review반영/
- Float, Integer -> primitive type으로 변경
- Lombok Annotation 적용
- Review.java : 변수 이름 변경 member -> user

chanwooo committed:
review반영/구현안된기능에 UnsupportedOperationException 적용

chanwooo committed:
review반영/Lombok Annotation 적용, default값 제거

reasonSun committed:
review반영/
- Lombok Annotation 적용
- image -> imageUrl

reasonSun commented:
TEST 코드 부분을 제외하고 DTO 관련 피드백 수정했습니다~!

reasonSun committed:
review반영/Entity Lombok Annotation 적용

f-lab-trey commented on src/main/java/flab/quing/review/Rating.java:
```diff
@@ -2,17 +2,21 @@
 import flab.quing.user.BaseEntity;
-import lombok.Getter;
-import lombok.Setter;
+import lombok.*;
```
Star import는 현업에서 잘 사용하지 않습니다. Intellij default 옵션이 켜져있어서 추가됐을 거 같은데 옵션 찾아서 꺼주시고 이 부분도 수정해주세요.

숙제) 왜 star import를 지양할까? 에 대해서 알아보기

f-lab-trey commented on src/main/java/flab/quing/review/Review.java:
```diff
@@ -3,22 +3,30 @@
 import flab.quing.user.User;
-import lombok.Getter;
-import lombok.Setter;
+import lombok.*;
```
ditto

f-lab-trey commented on src/main/java/flab/quing/review/Review.java:
```diff
@@ -3,22 +3,30 @@
 
+@Data
+@Setter(AccessLevel.NONE)
+@Builder
```
`@Builder` 대신 `@Accessors(chain=true)`를 달면, setter와 함께 builder 비슷한 코드를 작성할 수 있습니다. 이전 팀에서 그렇게 했었는데, `@Builder`에 이슈가 좀 있어서 그렇게 했다고 들었어요. 여유가 있으시면 한 번 찾아 보시고 저도 알려주세요.

`@Accessors(chain=true)` 관련 예
```
Rating rating = new Review().setUser().setRating();
```

f-lab-trey commented on src/main/java/flab/quing/review/dto/ReviewRequest.java:
```diff
@@ -1,10 +1,21 @@
-    private Long userId;
-    private Long waitingId;
-    private Integer rating;
+    @NonNull
```
primitive로 변경했으니까 `null`이 될 수 없겠죠? 그러면 이 어노테이션도 필요 없을 거 같네요.

f-lab-trey commented on src/main/java/flab/quing/review/dto/ReviewRequest.java:
```diff
@@ -1,10 +1,21 @@
+    @NonNull
+    private long userId;
+
+    @NonNull
```
ditto

f-lab-trey commented on src/main/java/flab/quing/store/dto/MenuRequest.java:
```diff
@@ -1,18 +1,22 @@
-@Getter
-@NoArgsConstructor
+@Data
+@Setter(AccessLevel.NONE)
```
`@Data` 안에 `@Setter`가 있을텐데, 한 번 더 사용하는 이유가 있을까요?

chanwooo replied:
`@Data` 안에 `@Setter`가 있지만 사용하지 못하도록 하고싶어서 의도적으로 동작하지않도록 하는 코드에요.

f-lab-trey commented on src/main/java/flab/quing/store/dto/MenuResponse.java:
```diff
@@ -1,19 +1,20 @@
 
-@Getter
+@Data
+@Setter(AccessLevel.NONE)
```
ditto

f-lab-trey commented on src/main/java/flab/quing/store/dto/StoreRequest.java:
```diff
@@ -1,11 +1,24 @@
+import lombok.Setter;
+
+@Data
+@Setter(AccessLevel.NONE)
```
ditto

f-lab-trey commented on src/main/java/flab/quing/user/dto/StoreMangerRequest.java:
```diff
@@ -1,7 +1,19 @@
+
     private String name;
+
     private String password;
```
[생각해보기]
Login시 암호화되지 않은 password가 request를 주고 받아도 괜찮을까요? 또한, password 같은 사용자의 민감한 정보를 DB에 저장할 때 암호화해서 저장하려면 어떻게 해야될까요?

f-lab-trey commented on src/main/java/flab/quing/store/dto/StoreResponse.java:
```diff
@@ -1,25 +1,26 @@
 
-@Getter
+@Data
+@Setter(AccessLevel.NONE)
```
ditto

f-lab-trey commented on src/main/java/flab/quing/user/dto/StoreManagerResponse.java:
```diff
@@ -1,19 +1,20 @@
 
-@Getter
+@Data
+@Setter(AccessLevel.NONE)
```
ditto

f-lab-trey commented on src/main/java/flab/quing/user/dto/StoreMangerRequest.java:
```diff
@@ -1,7 +1,19 @@
+import lombok.Setter;
+
+@Data
+@Setter(AccessLevel.NONE)
```
ditto

f-lab-trey commented on src/main/java/flab/quing/user/dto/UserRequest.java:
```diff
@@ -1,9 +1,18 @@
 
-@Getter
+@Data
+@Setter(AccessLevel.NONE)
```
ditto

f-lab-trey commented on src/main/java/flab/quing/user/dto/UserResponse.java:
```diff
@@ -1,17 +1,18 @@
 
-@Getter
+@Data
+@Setter(AccessLevel.NONE)
```
ditto

f-lab-trey commented on src/main/java/flab/quing/store/Menu.java:
```diff
@@ -1,20 +1,27 @@
 import javax.persistence.ManyToOne;
 
+@Data
+@Setter(AccessLevel.NONE)
```
앞서 남겼던 comment 참고해주세요.

f-lab-trey commented on src/main/java/flab/quing/store/Store.java:
```diff
@@ -1,16 +1,21 @@
 import javax.persistence.Entity;
 
+@Data
+@Setter(AccessLevel.NONE)
```
ditto

f-lab-trey commented on src/main/java/flab/quing/user/StoreManager.java:
```diff
@@ -1,4 +1,28 @@
+
+
+@Data
+@Setter(AccessLevel.NONE)
```
ditto

f-lab-trey commented on src/main/java/flab/quing/user/User.java:
```diff
@@ -1,18 +1,22 @@
 import javax.persistence.Entity;
 
+@Data
+@Setter(AccessLevel.NONE)
```
ditto

f-lab-trey commented on src/main/java/flab/quing/store/Menu.java:
```diff
@@ -1,20 +1,27 @@
 
+@Data
+@Setter(AccessLevel.NONE)
+@Builder
```
[질문]
Entity에 `@Builder`가 필요할까? 라는 의문이 있습니다. 테스트에서 사용하려고 추가하셨나요?

chanwooo replied:
여기엔 어떤게 필요하고 어떤건 안필요할까?에 대한 고민이 부족했던거같네요. 생성자만으로 충분할거같아요.

chanwooo replied:
좀 헷갈리네요.. builder보다 차라리 request - entity - response 간에 변환하는 메서드가 있으면 좋을거같단 생각이 들어요. 일단 builder가 있으면 범용적으로 쓸수있을거란 생각에 달은건데 고민을 좀 해보고 필요가 없으면 떼어내야할거같아요.

f-lab-trey replied:
네, 저도 나중에 필요할 때 붙이는 것이 조금 더 좋다고 생각합니다.

chanwooo committed:
feature/리뷰반영
- import * 제거
- Entity에 NoArgsConstructor access level protected 적용
- wrapped type 제거, @NonNull 제거

chanwooo committed:
Merge remote-tracking branch 'origin/feature/#12_서비스별_패키지_나누기' into feature/#12_서비스별_패키지_나누기

chanwooo committed:
feature/리뷰반영
- fix member -> user
- fix Integer -> int

f-lab-trey reviewed (approved):
LookGoodToMe!

f-lab-trey commented:
Merge 할 때는 squash and merge로 해주세요. 이유에 대해서는 과제로 조사 부탁드립니다.
